src/core.js

```diff
@@ -2386,119 +2414,167 @@ function FormatDateTime(dateTimeFormat, x) {
+                    }
+                    break;
+
+                  case 'timeZoneName':
+                    fv = ''; // TODO
+                    break;
 
                     // TODO: Era
-                    default:
-                        fv = tm['[['+ p +']]'];
+                  default:
+                    fv = tm['[['+ p +']]'];
                 }
+              }
+              // ix
+              arrPush.call(result, {
+                type: p,
+                value: fv
+              });
+            // f.
+            } else if (p === 'ampm') {
+              // i.
+              v = tm['[[hour]]'];
+              // ii./iii.
+              fv = resolveDateString(localeData, ca, 'dayPeriods', v > 11 ? 'pm' : 'am');
+              // iv.
+              arrPush.call(result, {
+                type: 'dayperiod',
+                value: fv
+              });
+            // g.
+            } else {
+              arrPush.call(result, {
+                type: 'separator',
+                value: pattern.substring(beginIndex, endIndex)
+              });
             }
-
-            // x. Replace the substring of result that consists of "{", p, and "}", with
-            //    fv.
-            result = result.replace('{'+ p +'}', fv);
+            // h.
+            index = endIndex + 1;
+            // i.
+            beginIndex = pattern.indexOf('{', index);
         }
-    }
-    // 8. If dateTimeFormat has an internal property [[hour12]] whose value is true, then
-    if (internal['[[hour12]]'] === true) {
-        // a. If pm is true, then let fv be an implementation and locale dependent String
-        //    value representing “post meridiem”; else let fv be an implementation and
-        //    locale dependent String value representing “ante meridiem”.
-        fv = resolveDateString(localeData, ca, 'dayPeriods', pm ? 'pm' : 'am');
-
-        // b. Replace the substring of result that consists of "{ampm}", with fv.
-        result = result.replace('{ampm}', fv);
-    }
+        // 12.
+        if (endIndex < pattern.length - 1) {
+          arrPush.call(result, {
+            type: 'separator',
+            value: pattern.substr(endIndex)
+          });
+        }
+        // 13.
+        return result;
+}
 
-    // Restore properties of the RegExp object
-    regexpState.exp.test(regexpState.input);
+/**
+ * When the FormatDateTime abstract operation is called with arguments dateTimeFormat
+ * (which must be an object initialized as a DateTimeFormat) and x (which must be a Number
+ * value), it returns a String value representing x (interpreted as a time value as
+ * specified in ES5, 15.9.1.1) according to the effective locale and the formatting
+ * options of dateTimeFormat.
+ */
+function FormatDateTime(dateTimeFormat, x) {
+  var parts = CreateDateTimeParts(dateTimeFormat, x);
+  var result = '';
 
-    // 9. Return result.
-    return result;
+  for (var part in parts) {
+      result += parts[part].value;
+  }
```

Comment:
this polyfill works in ES3 today :), in the future we should only care about es5 and above though.